Error parsing DateTime and Boolean fields on mutation

klaasvandeweerdt opened the pull request.

Commit:
#33782932 Error parsing datetime fields

Commit:
#33951246 Boolean `false` ignored in mutation

Commit:
#33951246 Boolean `false` ignored in mutation update

Commit:
Updated VCR_RECORD_MODE to `none`

klaasvandeweerdt commented:
@bennettgo , @den-sergeyev-workato 
I pushed another update. There was also an issue, in the same lambda, to include a boolean `false` value. A `false` value was excluded from the query, while `true` was included. I also updated the specs to include this specific scenario.

Commit:
#34352887 Incorrect character escaping in string values

klaasvandeweerdt commented:
@den-sergeyev-workato 
I added another fix for string parsing. It failed when using specific characters that require escaping in JSON.

This is an input example.
```
{
	"account": "wdc",
	"object": "requestUpdate{}",
	"argument_input": {
		"id": "Nl1...OTgy",
		"note": "**1\\. \" Bold**"
	}
}
```
I updated the code at line [2316](https://github.com/Workato-Partners/4me/blob/ba429b9bec381cff20ca4809a73510c0a172c054/connector.rb#L2316), changed it from `value = "\"#{value}\""` to `value = value.to_json` which solves the issue. I explicitly mention this as the same issue may exist in the generic GraphQL connector.

den-sergeyev-workato commented on connector.rb:
```diff
@@ -2307,12 +2307,13 @@
 
       # treat blank string for optional fields as nil
-      value = field['default'] if value.blank? && !optional
+      value = nil if value.is_a?(String) && optional && value == ''
```
Can you explain why it is not enough to have these 2 conditions?
```
value = nil if optional && value == ''
```

klaasvandeweerdt replied:
Optional string values should not be part of the GraphQL payload if they are empty, they get defaulted to `nil`. They are excluded in the next condition `unless value.nil? && optional`.

den-sergeyev-workato replied:
Got it, my question was are these two checks are equivalent? They are always false if value is not string equal to ""

```
value = nil if value.is_a?(String) && optional && value == ''
```
vs
```
value = nil if optional && value == ''
```

klaasvandeweerdt replied:
I copied it over from the universal GraphQL connector.
Updated it as you suggested as I cannot imagine a scenario where equal operator overload would return `''` for a non-string value.

den-sergeyev-workato commented on spec/actions/mutation_spec.rb:
```diff
@@ -28,4 +28,19 @@
+
+    it 'contains errors' do
+      expect(output).to include('errors')
+      puts output
```
Do we need puts here?

klaasvandeweerdt replied:
Sorry, still left from troubleshooting. I removed it.

den-sergeyev-workato commented on connector.rb:
```diff
@@ -2307,12 +2307,13 @@
+      unless value.nil? && optional
+        value = field['default'] if value.nil?
+
+        if %w[date_time date].include?(type) || (type == 'string' && control_type != 'select')
```
What changed so we change `ISO8601Timestamp, ISO8601DateTime, ISO8601Date` to `date_time date`?

klaasvandeweerdt replied:
A bug. `date` and `date_time` were not serialized properly (not encapsulated) as the comparison was wrong. `ISO8601Timestamp`, `ISO8601DateTime`, `ISO8601Date` are GraphQL types.

den-sergeyev-workato commented on connector.rb:
```diff
@@ -2307,12 +2307,13 @@
+      unless value.nil? && optional
+        value = field['default'] if value.nil?
+
+        if %w[date_time date].include?(type) || (type == 'string' && control_type != 'select')
```
What are possible options for control_type with type string?

klaasvandeweerdt replied:
If the type is a `select` then the value is an enum and should not be serialized.

den-sergeyev-workato replied:
Got it, I mean is it the only control_type we should not serialize? Just want to double check.

den-sergeyev-workato reviewed (commented):
Overall LGTM, some questions to clarify though.

I'm wondering why do we have so many changes in tapes, is it all affected by change?
from 
```"\"#{value}\""``` 
to 
```value.to_json```

Can you, please, link or description to first bug what this PR fixes?

klaasvandeweerdt commented:
@den-sergeyev-workato, there are 3 bug fixes.
1. The initial PR was for the incorrect parsing of `date_time` and `date` values, not encapsulated.
2. The `format_arg` lambda did not serialize a `false` value
3. The `format_arg` lambda did not properly escape special characters, changed it to `value.to_json` instead of `"\"#{value}\""`

I deleted the existing tape libraries, as all action tests were failing after the `value.to_json`. But also because our GraphQL API gets updated almost weekly and I wanted to make sure the recordings were updated.

Commit:
Implemented recommendation after review
